src/trpc/server/Params.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * @internal
+ */
+export interface Params<
+  TContextIn = unknown,
+  TContextOut = unknown,
+  TInputIn = unknown,
+  TInputOut = unknown,
+  TOutputIn = unknown,
+  TOutputOut = unknown,
+> {
+  /**
+   * @internal
+   */
+  _ctx_in: TContextIn;
+  /**
+   * @internal
+   */
+  _ctx_out: TContextOut;
+  /**
+   * @internal
+   */
+  _output_in: TOutputIn;
+  /**
+   * @internal
+   */
+  _output_out: TOutputOut;
+  /**
+   * @internal
+   */
+  _input_in: TInputIn;
+  /**
+   * @internal
+   */
+  _input_out: TInputOut;
+}
```

Comment:
@mmkal I guess this is what you referred to as an options-bag? :)

Reply:
Tried to write a reply to this on my phone but too hard! Yes this is pretty much what I was thinking. A couple of ways it could be simplified even further occur to me tho:

1. No linear generics at _all_. This might prove impossible if the compiler loses track of property types somewhere along the way, but here's a simplified example:

```ts
interface Params {
  input: unknown
  output: unknown
  context: unknown
}

const getClientThatUsesSpecificParams = <P extends Params>(params: P) => {
  return {
    useInput: (input: P['input']) => 123,
    getOutput: () => params.output,
  }
}

const specific = getClientThatUsesSpecificParams({
  input: {foo: 'abc'},
  output: [123, 456],
  context: {bar: 987},
})
```

TypeScript doesn't need specific typeargs for `TInput`, `TOutput`, `TContext` to keep track of them:

<img width="575" alt="image" src="https://user-images.githubusercontent.com/15040698/162837917-f9c4a1ad-15ce-4b1c-ab67-94eeb04a58d8.png">

2. Maybe there could be an IO type to generalise the `_in` and `_out` suffixes?

```ts
interface IO<I, O> {
  in: I
  out: O
}

interface Params< 
  TContextIn = unknown,
  TContextOut = unknown,
  TInputIn = unknown,
  TInputOut = unknown,
  TOutputIn = unknown,
  TOutputOut = unknown,
> {
  ctx: IO<TContextIn, TContextOut>
  input: IO<TInputIn, TInputOut>
  output: IO<TOutputIn, TOutputOut>
}
```

3. Both together!?!!?

```ts
interface IO {
  input: unknown
  output: unknown
}

interface Params {
  ctx: IO
  input: IO
  output: IO
}
```

Reply:
Made #35. Feel free to do PR, I have some decent testing and it's easy to run with yarn dev.

These are all internal types so easy to refactor later too, but would be nice to have it clean